firedrake/cython/dmcommon.pyx

```diff
@@ -523,6 +556,61 @@ cdef inline PetscInt _reorder_plex_closure(PETSc.DM dm,
         raise NotImplementedError(f"Not implemented for {dm.getCellType(p)}")
 
 
+def cell_orientation_from_vertex_arrangement(
+    cell_type,
+    vertex_ordering
+):
+    cdef:
+        PetscDMPolytopeType ct
+        PetscInt nvert, narr
+        PetscInt *arr
+
+    ct = cell_type
+    nvert = DMPolytopeTypeGetNumVertices(ct)
+    narr = DMPolytopeTypeGetNumArrangments(ct)
+    shift = narr // 2
+    for o in range(-shift, shift):
+        arr = DMPolytopeTypeGetVertexArrangment(ct, o)
+
+        # compare to vertex_ordering
+        equal = True
+        for i in range(nvert):
+            if arr[i] != vertex_ordering[i]:
+                equal = False
+                break
+        if equal:
+            return o
+
+    raise ValueError(f"No valid orientation found for vertex arrangement {vertex_ordering}")
+
+
+def _ordering(iterable):
+    iterable = list(iterable)
+    return tuple(iterable.index(p) for p in sorted(iterable))
+
+
+def orient_mesh(PETSc.DM dm, vertex_numbering):
+    cdef:
+        PetscDMPolytopeType ct
+        PetscInt o, nverts
+
+    pstart, pend = dm.getChart()
+    for p in range(pstart, pend):
+        cell_type = dm.getCellType(p)
+
+        # vertices have just a single orientation
+        if cell_type == PETSc.DM.PolytopeType.POINT:
+            continue
+
+        ct = cell_type
+        nverts = DMPolytopeTypeGetNumVertices(ct)
+        verts = dm.getTransitiveClosure(p)[0][-nverts:]
+        verts_renum = [vertex_numbering.getOffset(v) for v in verts]
+        vertex_ordering = _ordering(verts_renum)
+        o = cell_orientation_from_vertex_arrangement(cell_type, vertex_ordering)
+        DMPlexOrientPoint(dm.dm, p, o)
```

Comment:
To make this globally consistent, you need to use global section; see https://github.com/firedrakeproject/firedrake/blob/d28227133cb7bdd57f3da926b3e18c5ee638cb18/firedrake/mesh.py#L1092 and convert negative global number, corresponding to vertices owned by the other processes, using the `-(o + 1)` rule.

We decided to order DoFs on each entity relative to the cone of that entity for checkpointing purpose because the ordering of the cone is guaranteed to be preserved in the save-load cycle.
Modifying cones using `DMPlexOrientPoint()` based on the global vertex numbers is actually not desired as global vertex numbers are not preserved in the save-load cycle (or upon redistribution of the mesh).

I think we can make the global vertex numbers (and actually all the global entity numbers) preserved upon save/load or redistribution, but it has yet to be done. Even if it has been implemented, I think, we should abandon all approaches based on the global vertex numbers as hex requires a more generic approach anyway.

Reply:
> To make this globally consistent, you need to use global section

Sure. That should be straightforward to add.

> global vertex numbers is actually not desired as global vertex numbers are not preserved in the save-load cycle

I didn't realise this. How come checkpointing works currently then? What I have here is a simplification of the prior approach so the same limitations should apply.

> I think, we should abandon all approaches based on the global vertex numbers

I disagree with this. I think that we want to have this because it allows us to pack/unpack H(div) and H(curl) spaces without potentially expensive runtime transformations.

Reply:
Another contribution of this PR that I want to emphasise is that it [unifies the code paths](https://github.com/firedrakeproject/firedrake/pull/3332/files#diff-a0ea3805077584e4142e5bee782e15d2ed177efddaac05b34e339f2f8e1ebee4R654) for `cell_closure` and `entity_orientations`. I think we can pretty much rely on the orientations from `getTransitiveClosure` for our own work (with a [single transformation](https://github.com/firedrakeproject/firedrake/pull/3332/files#diff-a0ea3805077584e4142e5bee782e15d2ed177efddaac05b34e339f2f8e1ebee4R446) from canonical plex to canonical FIAT).

Basically PETSc tells us things like "relative to this cell, this edge is flipped". I believe that, subject to a canonical transformation, this is universal information applicable to both DMPlex and FIAT.

Reply:
> We decided to order DoFs on each entity relative to the cone of that entity for checkpointing purpose because the ordering of the cone is guaranteed to be preserved in the save-load cycle.
Modifying cones using DMPlexOrientPoint() based on the global vertex numbers is actually not desired as global vertex numbers are not preserved in the save-load cycle (or upon redistribution of the mesh).

I have thought more on this and this makes a lot of sense. The cones should not be arbitrarily modified upon loading the mesh. However, I still think that my approach has value. Could we introduce perhaps a `DMCopy` of the topology just for tabulating the closure, called, say, `closure_dm`? The advantage of having such a DM is then we get nicely ordered closures and can preserve the orientable properties of simplices and quads.